cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+      });
+
+      this->get_signals().post_advection_solver.connect([&](const SimulatorAccess<dim> &,
+                                                            const unsigned int ,
+                                                            const unsigned int ,
+                                                            const SolverControl &)
+      {
+        this->update();
+      });
+
+      n_material_data = material_file_names.size();
+      for (unsigned i = 0; i < n_material_data; i++)
+        {
+          if (material_file_format == perplex)
+            material_lookup.push_back(std::shared_ptr<MaterialUtilities::Lookup::MaterialLookup>
+                                      (new MaterialUtilities::Lookup::PerplexReader(datadirectory+material_file_names[i],
+                                                                                    /*use_bilinear_interpolation*/ true,
+                                                                                    this->get_mpi_communicator())));
+          else if (material_file_format == hefesto)
+            material_lookup.push_back(std::shared_ptr<MaterialUtilities::Lookup::MaterialLookup>
+                                      (new MaterialUtilities::Lookup::HeFESToReader(datadirectory+material_file_names[i],
+                                                                                    datadirectory+derivatives_file_names[i],
+                                                                                    /*use_bilinear_interpolation*/ true,
+                                                                                    this->get_mpi_communicator())));
+          else
+            AssertThrow (false, ExcNotImplemented());
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    EquilibriumGrainSize<dim>::update()
+    {
+      if (use_depth_dependent_viscosity)
+        {
+          std::vector<std::unique_ptr<internal::FunctorBase<dim>>> lateral_averaging_properties;
+          lateral_averaging_properties.push_back(std::make_unique<internal::FunctorDepthAverageUnscaledViscosity<dim>>());
+
+          std::vector<std::vector<double>> averages =
+            this->get_lateral_averaging().compute_lateral_averages(reference_viscosity_coordinates,
+                                                                   lateral_averaging_properties);
+
+          average_viscosity_profile.swap(averages[0]);
+
+          for (const auto &lateral_viscosity_average: average_viscosity_profile)
+            AssertThrow(numbers::is_finite(lateral_viscosity_average),
+                        ExcMessage("In computing depth averages, there is at"
+                                   " least one depth band that does not have"
+                                   " any quadrature points in it."
+                                   " Consider reducing number of depth layers"
+                                   " for averaging."));
+        }
+
+      initialized = true;
```

Comment:
Shouldn't this be part of the `initialize()` function?

Reply:
I didn't see this used anywhere in the code (we instead use `average_viscosity_profile.size() != 0` to check if the lateral average profile has been computed yet. So I think you can just remove this.